machineconfiguration/v1/types_machineconfignode.go

```diff
@@ -0,0 +1,266 @@
+	// The current version is the generation of the pinned image set that has most recently been successfully pulled and pinned on this node.
+	// The desired version is the generation of the pinned image set that is targeted to be pulled and pinned on this node.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=100
+	// +optional
+	PinnedImageSets []MachineConfigNodeStatusPinnedImageSet `json:"pinnedImageSets,omitempty"`
+}
+
+// MachineConfigNodeStatusPinnedImageSet holds information about the current and desired pinned image sets for the current observed machine config node.
+// +kubebuilder:validation:XValidation:rule="has(self.desiredGeneration) && has(self.currentGeneration) ? self.desiredGeneration >= self.currentGeneration : true",message="desired generation must be greater than or equal to the current generation"
+// +kubebuilder:validation:XValidation:rule="has(self.lastFailedGeneration) && has(self.desiredGeneration) ? self.desiredGeneration >= self.lastFailedGeneration : true",message="desired generation must be greater than last failed generation"
+// +kubebuilder:validation:XValidation:rule="has(self.lastFailedGeneration) ? has(self.desiredGeneration): true",message="desired generation must be defined if last failed generation is defined"
+type MachineConfigNodeStatusPinnedImageSet struct {
+	// name is the name of the pinned image set.
+	// Must be a lowercase RFC-1123 hostname (https://tools.ietf.org/html/rfc1123) consisting
+	// of only lowercase alphanumeric characters, hyphens (-), and periods (.), start and end
+	// with an alphanumeric character, and be at most 253 characters in length.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:XValidation:rule="!format.dns1123Subdomain().validate(self).hasValue()",message="a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character."
+	// +required
+	Name string `json:"name"`
+	// currentGeneration is the generation of the pinned image set that has most recently been successfully pulled and pinned on this node.
+	// +optional
+	CurrentGeneration int32 `json:"currentGeneration,omitempty"`
+	// desiredGeneration is the generation of the pinned image set that is targeted to be pulled and pinned on this node.
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	DesiredGeneration int32 `json:"desiredGeneration,omitempty"`
+	// lastFailedGeneration is the generation of the most recent pinned image set that failed to be pulled and pinned on this node.
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	LastFailedGeneration int32 `json:"lastFailedGeneration,omitempty"`
+	// lastFailedGenerationErrors is a list of errors why the lastFailed generation failed to be pulled and pinned.
+	// +kubebuilder:validation:MaxItems=10
+	// +optional
+	LastFailedGenerationErrors []string `json:"lastFailedGenerationErrors,omitempty"`
```

Comment:
This seems a somewhat weird way to keep track of a history of failures? I wonder if this should just be a singular error condition here